### PR TITLE
Appveyor support (fixes #859)

### LINKED ISSUE #859: Run stuff on appveyor
Because obviously creating virtualenvs on windows also needs a slight bit of testing

### PR DESCRIPTION
**Pull request by techtonik**
The result of test run - https://ci.appveyor.com/project/techtonik/virtualenv - this can be tuned later.

Pythons installed on Appveyor - http://www.appveyor.com/docs/installed-software#python

Further guide - http://python-packaging-user-guide.readthedocs.org/en/latest/appveyor/#adding-appveyor-support-to-your-project

### CODE REVIEW

**Comment by Ivoz** [comment]
Was wanting to copy matrix from [here](https://github.com/pypa/pip/blob/develop/appveyor.yml#L1-L10) for example

**Comment by techtonik** [comment]
Is https://github.com/pypa/pip/commit/15677747cf2977a652e2698535e7fd5f7689eb18 an AppVeyour bug or tox one? If the first, then it better be sent to @FeodorFitsner rather than worked around.

File: appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+install:
+  - cmd: python -m pip install tox
+
+build_script:
+  - cmd: python --version
+  - cmd: puthon -m tox
```
Comment (pfmoore) [inline appveyor.yml]: Typo. "puthon". But I agree with @Ivoz we want something more complete.

**Comment by pfmoore** [comment]
> Is pypa/pip@1567774 an AppVeyour bug or tox one

Neither. It's just a consequence of the available built in environments in tox. The change is fixing a bug in our appveyor.yml.

**Comment by techtonik** [comment]
I don't understand what exactly do you want. https://ci.appveyor.com/project/techtonik/virtualenv currently tests 5 python versions, all of which are failing for some reason that is beyond my knowledge.

**Comment by Ivoz** [comment]
@techtonik the matrix I suggested adds the python versions to appveyor, which is the reason they are failing.

**Comment by techtonik** [comment]
Now I see. `tox` actually doesn't find the Python versions. It is just misleading names in results that it reports.

**Comment by techtonik** [comment]
Fails on https://github.com/pypa/virtualenv/blob/352dcef7d9bdab298c4315bad1da564dd76901a5/tox.ini#L14 - `tox` uses `py` convention for current environment, and there is no such executable.
```
py runtests: commands[1] | python virtualenv.py C:\projects\virtualenv\.tox\py\tmp/venv 
Using real prefix 'C:\\Python27'
New python executable in C:\projects\virtualenv\.tox\py\tmp\venv\Scripts\python.EXE
Installing setuptools, pip, wheel...done.
py runtests: commands[2] | python virtualenv.py -p py C:\projects\virtualenv\.tox\py\tmp/py 
The executable py (from --python=py) does not exist
ERROR: InvocationError: 'C:\\projects\\virtualenv\\.tox\\py\\Scripts\\python.EXE virtualenv.py -p py C:\\projects\\virtualenv\\.tox\\py\\tmp/py' 
___________________________________ summary ___________________________________
ERROR:   py: commands failed
Command exited with code 1
```

**Comment by Ivoz** [comment]
I would try adding `TOXENV`'s back in, with their value being the name of the binary.

e.g

    environment:
      matrix:
        - PYTHON: "C:\\Python27"
        - TOXENV: "python2.7"

&

    test_script:
      - "%PYTHON%\\python.exe -m tox -e %TOXENV%"

**Comment by techtonik** [comment]
@Ivoz that won't work, because there is no such binaries on Windows.
```
>virtualenv -p python3.5 asdf
The executable python3.5 (from --python=python3.5) does not exist

>py -3
Python 3.5.1 (v3.5.1:37a07cee5969, Dec  6 2015, 01:38:48) [MSC v.1900 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>>
```

The line `python virtualenv.py -p {envname} {envtmpdir}/{envname}` is a hack that needs to be removed for it to work.

**Comment by pfmoore** [comment]
The ```py``` environment should work (and doesn't need to be added to envlist as it's built in) but our tests are buggy. Specifically

```
commands =
    py.test
    python virtualenv.py {envtmpdir}/venv
    python virtualenv.py -p {envname} {envtmpdir}/{envname}
    python virtualenv.py --python={envpython} {envtmpdir}/{envname}-fullpath
```

The 3 "create a virtualenv" tests have a number of problems. They should really be rewritten as proper py.test tests.

Issues include:

1. Using ```python``` as the command rather than ```{envpython}```
2. The third test uses ```--python={envpython}``` but tox omits the ```.exe``` from the command on Windows, whereas ```virtualenv``` needs the ```.exe``` to be included in the ```--python``` argument.
3. The ```-p {envname}``` test relies on the envname being a valid short form for the ```-p``` option. The envnames we use are valid short forms on Unix but not on Windows (which uses ```pyXY```), and anyway we want to use an envname of ```py``` on Appveyor.

I'll work up a PR making these tests a proper part of the test suite.

**Comment by pfmoore** [comment]
OK, #863 committed to fix the test issues.

**Comment by techtonik** [comment]
It can not pass - py26 does not exist in Windows - https://ci.appveyor.com/project/techtonik/virtualenv/build/job/2ljpd1ydxhvpjsf0

**Comment by pfmoore** [comment]
It needs a few more fixes (and there is an issue that the tests on Windows exposed). I have a PR in progress that includes ```appveyor.yml``` and fixes up the tests. See https://ci.appveyor.com/project/pfmoore/virtualenv/build/1.0.23 which is the latest (hopefully final!) iteration.

(The correct invocation for the short form on Windows is ```-p 2.6```, not ```-p py26```, which is not documented anywhere. The way the original tests checked this was Unix-specific and very fragile).

**Comment by pfmoore** [comment]
If my branch gets all green on Appveyor, I'll make a PR from it, and merge that in preference to this one. Hope that's OK - your work in putting together this PR is much appreciated, as without it I'd have never got round to working on this.

**Comment by pfmoore** [comment]
OK, #865 has been merged, which adds appveyor support based on the ideas here. Many thanks @techtonik for the contribution.

**Comment by techtonik** [comment]
Ugh, okay. But thanks are not as easy to add to resume as https://github.com/pypa/virtualenv/graphs/contributors

**Pull request closed without merging**